src/Fuzz.elm

```diff
@@ -217,122 +241,538 @@ charGenerator =
+    map (\( a, b, c, d, e ) -> transform a b c d e) (tuple5 ( fuzzA, fuzzB, fuzzC, fuzzD, fuzzE ))
+
+
+{-| Map over many fuzzers. This can act as mapN for N > 5.
+-}
+andMap : Fuzzer (a -> b) -> Fuzzer a -> Fuzzer b
+andMap =
+    map2 (<|)
+
+
+{-| Create a fuzzer based on the result of another fuzzer.
+-}
+andThen : (a -> Fuzzer b) -> Fuzzer a -> Fuzzer b
+andThen transform (Internal.Fuzzer baseFuzzer) =
+    Internal.Fuzzer
+        (\noShrink ->
+            case baseFuzzer noShrink of
+                Gen genVal ->
+                    Gen <| Random.andThen genVal (transform >> Internal.unpackGenVal)
+
+                Shrink genTree ->
+                    Shrink <| andThenRoseTrees transform genTree
+        )
+
+
+andThenRoseTrees : (a -> Fuzzer b) -> Generator (RoseTree a) -> Generator (RoseTree b)
+andThenRoseTrees transform genTree =
+    Random.andThen genTree
+        (\(Rose root branches) ->
+            let
+                -- genOtherChildren : Generator (LazyList (RoseTree b))
+                genOtherChildren =
+                    branches
+                        |> Lazy.List.map (\rt -> RoseTree.map (transform >> Internal.unpackGenTree) rt |> unwindRoseTree)
+                        |> unwindLazyList
+                        |> Random.map (Lazy.List.map RoseTree.flatten)
+            in
+                Random.map2
+                    (\(Rose trueRoot root'sChildren) otherChildren ->
+                        Rose trueRoot (Lazy.List.append root'sChildren otherChildren)
+                    )
+                    (Internal.unpackGenTree (transform root))
+                    genOtherChildren
+        )
+
+
+unwindRoseTree : RoseTree (Generator a) -> Generator (RoseTree a)
+unwindRoseTree (Rose genRoot lazyListOfRoseTreesOfGenerators) =
+    case Lazy.List.headAndTail lazyListOfRoseTreesOfGenerators of
+        Nothing ->
+            Random.map RoseTree.singleton genRoot
+
+        Just ( Rose gen children, moreList ) ->
+            Random.map4 (\a b c d -> Rose a (Lazy.List.cons (Rose b c) d))
+                genRoot
+                gen
+                (Lazy.List.map unwindRoseTree children |> unwindLazyList)
+                (Lazy.List.map unwindRoseTree moreList |> unwindLazyList)
+
+
+unwindLazyList : LazyList (Generator a) -> Generator (LazyList a)
+unwindLazyList lazyListOfGenerators =
+    case Lazy.List.headAndTail lazyListOfGenerators of
+        Nothing ->
+            Random.constant Lazy.List.empty
+
+        Just ( head, tail ) ->
+            Random.map2 Lazy.List.cons head (unwindLazyList tail)
+
+
+{-| Filter values from a fuzzer that do not satisfy the provided predicate.
+Shrunken values will also satisfy the predicate, however a restrictive predicate
+will make shrinking less effective.
+
+**Warning:** The predicate must return `True` fairly often for values generated
+by the input fuzzer. If the predicate rarely or never returns `True`, using the
+resulting fuzzer will crash your program.
```

Comment:
It would be awesome to include that explanation in the docs...when I read it my first impression was "whoa that sounds like a nasty bug" - it did not occur to me that it was innate to the semantics of `filter` in the context of randomly generating values.